In [1]:
import sqlalchemy
%load_ext sql
%sql mysql+pymysql://student:mds-is-fun@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
%sql student@witchcraft

### 1. Getting to know the data
#### a) Show a list of the tables included in the database.

In [2]:
%%sql
SHOW TABLES

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
37 rows affected.


[('accused',),
 ('accused_family',),
 ('appeal',),
 ('calendarcustom',),
 ('case_person',),
 ('commission',),
 ('complaint',),
 ('confession',),
 ('counterstrategy',),
 ('demonicpact',),
 ('denunciation',),
 ('devilappearance',),
 ('elf_fairyelements',),
 ('imprisonment',),
 ('linkedtrial',),
 ('malice',),
 ('mentionedaswitch',),
 ('movestohla',),
 ('musicalinstrument',),
 ('ordeal',),
 ('othercharges',),
 ('othernamedwitch',),
 ('person',),
 ('prevcommission',),
 ('propertydamage',),
 ('ref_parish',),
 ('reference',),
 ('religiousmotif',),
 ('ritualobject',),
 ('shapechanging',),
 ('source',),
 ('torture',),
 ('trial',),
 ('trial_person',),
 ('weathermodification',),
 ('whitemagic',),
 ('witchesmeetingplace',)]

#### b)Display the column names for the table accused.

In [4]:
%%sql
DESCRIBE accused

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
31 rows affected.


[('accusedref', 'varchar(10)', 'YES', '', None, ''),
 ('accusedsystemid', 'varchar(3)', 'YES', '', None, ''),
 ('accusedid', 'double', 'YES', '', None, ''),
 ('firstname', 'varchar(36)', 'YES', '', None, ''),
 ('lastname', 'varchar(25)', 'YES', '', None, ''),
 ('m_firstname', 'varchar(29)', 'YES', '', None, ''),
 ('m_surname', 'varchar(21)', 'YES', '', None, ''),
 ('alias', 'varchar(26)', 'YES', '', None, ''),
 ('patronymic', 'varchar(14)', 'YES', '', None, ''),
 ('destitle', 'varchar(28)', 'YES', '', None, ''),
 ('sex', 'varchar(6)', 'YES', '', None, ''),
 ('age', 'double', 'YES', '', None, ''),
 ('age_estcareer', 'tinyint', 'YES', '', None, ''),
 ('age_estchild', 'tinyint', 'YES', '', None, ''),
 ('res_settlement', 'varchar(22)', 'YES', '', None, ''),
 ('res_parish', 'varchar(10)', 'YES', '', None, ''),
 ('res_presbytery', 'varchar(17)', 'YES', '', None, ''),
 ('res_county', 'varchar(13)', 'YES', '', None, ''),
 ('res_burgh', 'varchar(13)', 'YES', '', None, ''),
 ('res_ngr_letters', 'varchar(2)', 'YES', '', None, ''),
 ('res_ngr_easting', 'double', 'YES', '', None, ''),
 ('res_ngr_northing', 'double', 'YES', '', None, ''),
 ('ethnic_origin', 'varchar(7)', 'YES', '', None, ''),
 ('maritalstatus', 'varchar(15)', 'YES', '', None, ''),
 ('socioecstatus', 'varchar(15)', 'YES', '', None, ''),
 ('occupation', 'varchar(14)', 'YES', '', None, ''),
 ('notes', 'text', 'YES', '', None, ''),
 ('createdby', 'varchar(3)', 'YES', '', None, ''),
 ('createdate', 'datetime', 'YES', '', None, ''),
 ('lastupdatedby', 'varchar(3)', 'YES', '', None, ''),
 ('lastupdatedon', 'datetime', 'YES', '', None, '')]

#### c) How many people are included in the accused table?

In [8]:
%%sql
SELECT COUNT(DISTINCT accusedid)
FROM accused

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
1 rows affected.


[(3006,)]

#### d) Display the columns firstname, sex, and age for 5 cases in the accused table.

In [9]:
%%sql
SELECT firstname, sex,age 
FROM accused
LIMIT 5

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
5 rows affected.


[('Mareon', 'Female', None),
 ('Thom', 'Male', None),
 ('Christian', 'Female', None),
 ('Janet', 'Female', None),
 ('Agnes', 'Female', None)]

#### e)  Looks like the age is missing for some observations. Count the number of nonmissing values for age in the data.

In [10]:
%%sql
SELECT COUNT(age)
FROM accused
WHERE age != "None"

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
1 rows affected.


[(166,)]

#### f) Show a list of unique occupations.

In [12]:
%%sql
SELECT DISTINCT occupation 
FROM accused

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
32 rows affected.


[(None,),
 ('Servant',),
 ('Vagabond',),
 ('Weaver',),
 ('Midwife',),
 ('Tailor',),
 ('Messenger',),
 ('Brewster',),
 ('Smith',),
 ('Minister',),
 ('Farmer',),
 ('Mealmaker',),
 ('Tasker',),
 ('Maltman',),
 ('Miller',),
 ('Healer',),
 ('Loadman',),
 ('Blacksmith',),
 ('Shop-keeper',),
 ('Stabler',),
 ('Nurse',),
 ('Collier',),
 ('School teacher',),
 ('Slaterer',),
 ('Workman',),
 ('School Master',),
 ('Merchant',),
 ('Fisherman',),
 ('Henwife',),
 ('Cook',),
 ('Sailor',),
 ('Creelman',)]

### 2. Seeing the Devil in `devilappearance` table.
#### a)List the unique devil_type in the data.

In [18]:
%%sql
DESCRIBE devilappearance

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
10 rows affected.


[('devilref', 'varchar(10)', 'YES', '', None, ''),
 ('devilsystemid', 'varchar(2)', 'YES', '', None, ''),
 ('devilid', 'double', 'YES', '', None, ''),
 ('caseref', 'varchar(10)', 'YES', '', None, ''),
 ('devil_type', 'varchar(22)', 'YES', '', None, ''),
 ('devil_text', 'varchar(50)', 'YES', '', None, ''),
 ('createdby', 'varchar(3)', 'YES', '', None, ''),
 ('createdate', 'datetime', 'YES', '', None, ''),
 ('lastupdatedby', 'varchar(3)', 'YES', '', None, ''),
 ('lastupdatedon', 'datetime', 'YES', '', None, '')]

In [14]:
%%sql
SELECT DISTINCT devil_type
FROM devilappearance

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
16 rows affected.


[('Male',),
 ('Female Fairy',),
 ('Male Fairy',),
 ('Animal Devil',),
 ('Spirit',),
 ('Ghost',),
 ('Other Demon',),
 ('Female',),
 (None,),
 ('Fairy',),
 ('Unspecified Devil',),
 ('Insect Devil',),
 ('Baby',),
 ('Human Devil',),
 ('Child Devil',),
 ('Inanimate Object Devil',)]

#### b) There is also a little description of the type of the devil sighting in the devil_text column. How many of the sightings mention the word "black" in the description?

In [17]:
%%sql
SELECT COUNT(devil_text)
FROM devilappearance
WHERE devil_text LIKE "%black%"

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
1 rows affected.


[(121,)]

#### c) What proportion of the devils (in devil_type) are male?

In [21]:
%%sql 
SELECT 
    DISTINCT devil_type,
    (COUNT(devil_type) / (SELECT COUNT(devilref) FROM devilappearance)) as Proportation
FROM devilappearance
WHERE devil_type='male'
GROUP BY devil_type

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
1 rows affected.


[('Male', Decimal('0.6313'))]

### 3.The trial
#### a) What are the average and maximum numbers of male and female accusers?

In [22]:
%%sql
DESCRIBE trial

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
65 rows affected.


[('trialref', 'varchar(9)', 'YES', '', None, ''),
 ('trialid', 'double', 'YES', '', None, ''),
 ('trialsystemid', 'varchar(2)', 'YES', '', None, ''),
 ('caseref', 'varchar(10)', 'YES', '', None, ''),
 ('trialtype', 'double', 'YES', '', None, ''),
 ('trial_settlement', 'varchar(11)', 'YES', '', None, ''),
 ('trial_parish', 'varchar(9)', 'YES', '', None, ''),
 ('trial_presbytery', 'varchar(11)', 'YES', '', None, ''),
 ('trial_county', 'varchar(10)', 'YES', '', None, ''),
 ('trial_burgh', 'varchar(13)', 'YES', '', None, ''),
 ('trial_ngr_letters', 'tinyint', 'YES', '', None, ''),
 ('trial_ngr_easting', 'tinyint', 'YES', '', None, ''),
 ('trial_ngr_northing', 'tinyint', 'YES', '', None, ''),
 ('watching', 'tinyint', 'YES', '', None, ''),
 ('watchingdate', 'tinyint', 'YES', '', None, ''),
 ('watchingdate_as_date', 'tinyint', 'YES', '', None, ''),
 ('arrest', 'tinyint', 'YES', '', None, ''),
 ('arrestdate', 'varchar(10)', 'YES', '', None, ''),
 ('arrestdate_as_date', 'datetime', 'YES', '', None, ''),
 ('fled', 'tinyint', 'YES', '', None, ''),
 ('fleddate', 'varchar(10)', 'YES', '', None, ''),
 ('fleddate_as_date', 'datetime', 'YES', '', None, ''),
 ('actiondropped', 'tinyint', 'YES', '', None, ''),
 ('actiondropdate', 'varchar(10)', 'YES', '', None, ''),
 ('actiondropdate_as_date', 'datetime', 'YES', '', None, ''),
 ('actiondroppedby', 'tinyint', 'YES', '', None, ''),
 ('actionbody', 'varchar(20)', 'YES', '', None, ''),
 ('confrontingsuspects', 'tinyint', 'YES', '', None, ''),
 ('pretrialnotes', 'text', 'YES', '', None, ''),
 ('processtype', 'tinyint', 'YES', '', None, ''),
 ('trialdate', 'varchar(11)', 'YES', '', None, ''),
 ('trialdate_as_date', 'datetime', 'YES', '', None, ''),
 ('trialplace', 'varchar(10)', 'YES', '', None, ''),
 ('female_accusers', 'double', 'YES', '', None, ''),
 ('male_accusers', 'double', 'YES', '', None, ''),
 ('high_status', 'tinyint', 'YES', '', None, ''),
 ('defence', 'tinyint', 'YES', '', None, ''),
 ('verdict', 'varchar(11)', 'YES', '', None, ''),
 ('sentence', 'varchar(18)', 'YES', '', None, ''),
 ('cjtorder', 'tinyint', 'YES', '', None, ''),
 ('cjtdate', 'date', 'YES', '', None, ''),
 ('cjtdate_as_date', 'datetime', 'YES', '', None, ''),
 ('circuit', 'tinyint', 'YES', '', None, ''),
 ('circuitname', 'varchar(14)', 'YES', '', None, ''),
 ('localwithcrep', 'tinyint', 'YES', '', None, ''),
 ('noreftocentral', 'tinyint', 'YES', '', None, ''),
 ('trialnotes', 'text', 'YES', '', None, ''),
 ('execution', 'tinyint', 'YES', '', None, ''),
 ('executionmethod', 'varchar(15)', 'YES', '', None, ''),
 ('executiondate', 'varchar(10)', 'YES', '', None, ''),
 ('executiondate_as_date', 'datetime', 'YES', '', None, ''),
 ('executionplace', 'varchar(36)', 'YES', '', None, ''),
 ('exec_settlement', 'tinyint', 'YES', '', None, ''),
 ('exec_parish', 'varchar(9)', 'YES', '', None, ''),
 ('exec_presbytery', 'tinyint', 'YES', '', None, ''),
 ('exec_county', 'varchar(10)', 'YES', '', None, ''),
 ('exec_burgh', 'varchar(11)', 'YES', '', None, ''),
 ('exec_ngr_letters', 'tinyint', 'YES', '', None, ''),
 ('exec_ngr_easting', 'tinyint', 'YES', '', None, ''),
 ('exec_ngr_northing', 'tinyint', 'YES', '', None, ''),
 ('posttrialnotes', 'text', 'YES', '', None, ''),
 ('createdby', 'varchar(3)', 'YES', '', None, ''),
 ('createdate', 'datetime', 'YES', '', None, ''),
 ('lastupdatedby', 'varchar(3)', 'YES', '', None, ''),
 ('lastupdatedon', 'datetime', 'YES', '', None, '')]

In [24]:
%%sql
SELECT
ROUND(AVG(female_accusers),2) as avg_female_accusers,
ROUND(MAX(female_accusers),2) as max_female_accusers,
ROUND(AVG(male_accusers),2) as avg_male_accusers,
ROUND(MAX(male_accusers),2) as max_male_accusers
FROM trial

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
1 rows affected.


[(0.26, 27.0, 0.42, 48.0)]

#### b) Count the number of sentences by sentence type. List them in a table (in descending order), excluding missing values. Rename the column headings to something sensible.

In [31]:
%%sql
SELECT sentence, COUNT(*) as num_sec
FROM trial
WHERE sentence IS NOT NULL
GROUP BY sentence
ORDER BY 2 DESC

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
10 rows affected.


[('Execution', 205),
 ('Released', 52),
 ('Banishment', 27),
 ('Declared Fugitive', 11),
 ('Excommunicated', 6),
 ('Put to the horn', 2),
 ('Hang', 1),
 ('Branded', 1),
 ('Prison', 1),
 ('Public Humiliation', 1)]

#### c)Do the number of accusers matter for the verdict? Compare the average number of accusers by the type of verdict. Again make sure the table is sorted and the headings make sense.

In [34]:
%%sql
SELECT verdict,
ROUND(AVG(female_accusers+male_accusers),3) as avg_accuse
FROM trial
GROUP BY verdict
ORDER BY 2 DESC

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
5 rows affected.


[('Guilty', 5.804),
 ('Not Guilty', 4.558),
 ('Not Proven', 1.0),
 ('Half Guilty', 0.286),
 (None, 0.201)]

### 4.Tortured Truth
#### a) Left join the trial and confession tables. For what share of trials does the database record confessions? Create a results table with the number of all trials, the number of confessions, and the share of trials with confessions recorded.

In [35]:
%%sql
DESCRIBE confession

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
18 rows affected.


[('confessionref', 'varchar(9)', 'YES', '', None, ''),
 ('confessionsystemid', 'varchar(2)', 'YES', '', None, ''),
 ('confessionid', 'double', 'YES', '', None, ''),
 ('trialref', 'varchar(9)', 'YES', '', None, ''),
 ('centraltrialconfession', 'tinyint', 'YES', '', None, ''),
 ('confessionrec', 'tinyint', 'YES', '', None, ''),
 ('confessiondate', 'varchar(16)', 'YES', '', None, ''),
 ('confessiondate_as_date', 'datetime', 'YES', '', None, ''),
 ('confessionplace', 'varchar(14)', 'YES', '', None, ''),
 ('confessionlocation', 'varchar(21)', 'YES', '', None, ''),
 ('confessiontext', 'tinyint', 'YES', '', None, ''),
 ('confessiondetails', 'tinyint', 'YES', '', None, ''),
 ('confessionret', 'tinyint', 'YES', '', None, ''),
 ('notes', 'tinyint', 'YES', '', None, ''),
 ('createdby', 'varchar(3)', 'YES', '', None, ''),
 ('createdate', 'datetime', 'YES', '', None, ''),
 ('lastupdatedby', 'varchar(3)', 'YES', '', None, ''),
 ('lastupdatedon', 'datetime', 'YES', '', None, '')]

In [40]:
%%sql 
SELECT
COUNT(DISTINCT(t.trialref)) as num_trial,
COUNT(DISTINCT(c.trialref)) as num_conf,
(COUNT(DISTINCT(c.trialref))/COUNT(DISTINCT(t.trialref))) as trials_with_confessions
FROM trial t
LEFT JOIN confession c
ON t.trialref=c.trialref

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
1 rows affected.


[(3211, 756, Decimal('0.2354'))]

#### b) Only a small number of trials have records of torture. Is there a higher share of confessions among trials with records of torture than trials without such record? Hint: You will need to merge on the confession table.

In [42]:
%%sql 
DESCRIBE torture

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
12 rows affected.


[('tortureref', 'varchar(9)', 'YES', '', None, ''),
 ('torturesystemid', 'varchar(2)', 'YES', '', None, ''),
 ('tortureid', 'double', 'YES', '', None, ''),
 ('trialref', 'varchar(9)', 'YES', '', None, ''),
 ('tortureexists', 'tinyint', 'YES', '', None, ''),
 ('torturedate', 'varchar(7)', 'YES', '', None, ''),
 ('torturedate_as_date', 'datetime', 'YES', '', None, ''),
 ('torturetype', 'varchar(19)', 'YES', '', None, ''),
 ('createdby', 'varchar(3)', 'YES', '', None, ''),
 ('createdate', 'datetime', 'YES', '', None, ''),
 ('lastupdatedby', 'varchar(3)', 'YES', '', None, ''),
 ('lastupdatedon', 'datetime', 'YES', '', None, '')]

In [48]:
%%sql

SELECT COUNT(c.trialref)/COUNT(t.trialref) trials_with_confessions
FROM trial t
LEFT JOIN confession c ON t.trialref = c.trialref
LEFT JOIN torture ON c.trialref = torture.trialref
WHERE tortureref IS NOT NULL

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
1 rows affected.


[(Decimal('1.0000'),)]

In [47]:
%%sql

SELECT COUNT(c.trialref)/COUNT(t.trialref) trials_with_confessions
FROM trial t
LEFT JOIN confession c ON t.trialref = c.trialref
LEFT JOIN torture ON c.trialref = torture.trialref
WHERE tortureref IS NULL

 * mysql+pymysql://student:***@witchcraft.csbmzoea3lu9.us-east-1.rds.amazonaws.com/witchcraft
1 rows affected.


[(Decimal('0.2696'),)]

The result suggests that, the share of trials with the record of torture is 100%, while the share of confessions without torture is 26.96%. Which means the assumption is correct.